src/libraries/Common/src/System/Reflection/TypeNameParser.Helpers.cs

```diff
@@ -0,0 +1,195 @@
+                    separator--;
+                typeNamespace = typeName.Substring(0, separator);
+                name = typeName.Substring(separator + 1);
+            }
+
+            return (typeNamespace, name);
+        }
+
+        private Type? Resolve(Metadata.TypeName typeName)
+        {
+            if (typeName.IsNested)
+            {
+                Metadata.TypeName? current = typeName;
+                int nestingDepth = 0;
+                while (current is not null && current.IsNested)
+                {
+                    nestingDepth++;
+                    current = current.DeclaringType;
+                }
+
+                // We're performing real type resolution, it is assumed that the caller has already validated the correctness
+                // of this TypeName object against their own policies, so there is no need for this method to perform any further checks.
+                string[] nestedTypeNames = new string[nestingDepth];
+                current = typeName;
+                while (current is not null && current.IsNested)
+                {
+#if MONO
+                    nestedTypeNames[--nestingDepth] = current.Name;
+#else // CLR, NativeAOT and tools require unescaped nested type names
+                    nestedTypeNames[--nestingDepth] = UnescapeTypeName(current.Name);
+#endif
+                    current = current.DeclaringType;
+                }
+#if SYSTEM_PRIVATE_CORELIB
+                string nonNestedParentName = current!.FullName;
+#else // the tools require unescaped names
+                string nonNestedParentName = UnescapeTypeName(current!.FullName);
+#endif
+
+                Type? type = GetType(nonNestedParentName, nestedTypeNames, typeName.AssemblyName, typeName.FullName);
+                return Make(type, typeName);
+            }
+            else if (typeName.IsConstructedGenericType)
+            {
+                return Make(Resolve(typeName.GetGenericTypeDefinition()), typeName);
+            }
+            else if (typeName.IsArray || typeName.IsPointer || typeName.IsByRef)
+            {
+                Metadata.TypeName elementType = typeName.GetElementType();
+
+                if (elementType.IsByRef || (typeName.IsVariableBoundArrayType && typeName.GetArrayRank() > 32))
+                {
+#if SYSTEM_PRIVATE_CORELIB
+                    throw new TypeLoadException(); // CLR throws TypeLoadException for invalid decorators
+#else
+                    return null;
+#endif
+                }
+
+                return Make(Resolve(elementType), typeName);
+            }
+            else
+            {
+                Type? type = GetType(
+#if SYSTEM_PRIVATE_CORELIB
+                    typeName.FullName,
+#else // the tools require unescaped names
+                    UnescapeTypeName(typeName.FullName),
+#endif
+                    nestedTypeNames: ReadOnlySpan<string>.Empty, typeName.AssemblyName, typeName.FullName);
+
+                return Make(type, typeName);
```

Comment:
I do not see why we call Make here. Can we just return `type`?

Reply:
Or why do we need Make method at all - it does nearly the same sequence of if checks as this method. Can we just call the Make methods here directly?

Reply:
> Or why do we need Make method at all

It was caused by the fact that if given type name is both nested and generic, it should be solved as a nested type first and then turned into generic. I've simplified the code so it's clear now.